core/store/src/cold_storage.rs

```diff
@@ -0,0 +1,303 @@
+/// Rather than manually combining those parts in the right order for every cold column,
+/// we define `DBCol::key_type` to determine how a key for the column is formed,
+/// `get_keys_from_store` to determine all possible keys only for needed key parts,
+/// and `combine_keys` to generated all possible whole keys for the column based on order of those parts.
+///
+/// To add a new column to cold storage, we need to
+/// 1. add it to `DBCol::is_cold` list
+/// 2. define `DBCol::key_type` for it (if it isn't already defined)
+/// 3. add new clause in `get_keys_from_store` for new key types used for this column (if there are any)
+pub fn update_cold_db(
+    cold_db: &dyn Database,
+    hot_store: &Store,
+    height: &BlockHeight,
+) -> io::Result<()> {
+    let _span = tracing::debug_span!(target: "store", "update cold db", height = height);
+
+    let mut store_with_cache = StoreWithCache { store: hot_store, cache: StoreCache::new() };
+
+    let key_type_to_keys = get_keys_from_store(&mut store_with_cache, height)?;
+    for col in DBCol::iter() {
+        if col.is_cold() {
+            copy_from_store(
+                cold_db,
+                &mut store_with_cache,
+                col,
+                combine_keys(&key_type_to_keys, &col.key_type()),
+            )?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Gets values for given keys in a column from provided hot_store.
+/// Creates a transaction based on that values with set DBOp s.
+/// Writes that transaction to cold_db.
+fn copy_from_store(
+    cold_db: &dyn Database,
+    hot_store: &mut StoreWithCache,
+    col: DBCol,
+    keys: Vec<StoreKey>,
+) -> io::Result<()> {
+    let _span = tracing::debug_span!(target: "store", "create and write transaction to cold db", col = %col);
+
+    let mut transaction = DBTransaction::new();
+    for key in keys {
+        // TODO: Look into using RocksDB’s multi_key function.  It
+        // might speed things up.  Currently our Database abstraction
+        // doesn’t offer interface for it so that would need to be
+        // added.
+        let data = hot_store.get(col, &key)?;
+        if let Some(value) = data {
+            // Database checks col.is_rc() on read and write
+            // And in every way expects rc columns to be written with rc
+            //
+            // TODO: As an optimisation, we might consider breaking the
+            // abstraction layer.  Since we’re always writing to cold database,
+            // rather than using `cold_db: &dyn Database` argument we cloud have
+            // `cold_db: &ColdDB` and then some custom function which lets us
+            // write raw bytes.
+            if col.is_rc() {
+                transaction.update_refcount(
+                    col,
+                    key,
+                    add_positive_refcount(&value, std::num::NonZeroU32::new(1).unwrap()),
+                );
+            } else {
+                transaction.set(col, key, value);
+            }
```

Comment:
One idea (but perhaps for the future) might be to pass `&ColdDB` to the function rather than operating on `&dyn Database`. We could then have a custom write function which would just store raw bytes. This would allow us to avoid some of the data manipulation.